### ISSUE TITLE
Optional.unwrap() in winpcap causing crash

### ISSUE THREAD

**Issue by bgourlie** [libpnet/libpnet#281]
https://github.com/libpnet/libpnet/blob/47f48e00282aabc005c96393930cb3408139a7f5/src/datalink/winpcap.rs#L277

I can try my hand at a fix, but I'm not sure how to handle the case where the option is Nothing.

**Comment by mrmonday** [comment]
If you take a look at the preceding code, you'll notice there's an if statement to handle the case where are no packets in the buffer. Rather than attempting to handle the error case in this line, I would identify the issue with the preceding code causing the buffer to be empty, and handle the error there.

I'm happy to provide more guidance for this if you would like - I don't have a Windows box to test on right now, and it all worked last time I tried it (and, indeed, it works on CI).

**Comment by bgourlie** [comment]
Yeah, interestingly, it successfully reads UDP packets for some time before crashing, so it's not as if its not working at all. I'll take a look at this later tonight and see if I can get a fix submitted.

**Comment by bgourlie** [comment]
I've now seen this error out earlier in the function as well:

https://github.com/libpnet/libpnet/blob/47f48e00282aabc005c96393930cb3408139a7f5/src/datalink/winpcap.rs#L261

The error message, which is not verbatim since I forgot to copy it, was: "The operation completed successfully" and the OS return code was 0.

**Comment by mrmonday** [comment]
These might help your investigation:
 * https://github.com/patmarion/winpcap/blob/master/WpcapSrc_4_1_3/Packet9x/DLL/Packet32.c#L524-L533
 * https://github.com/patmarion/winpcap/blob/master/WpcapSrc_4_1_3/Packet9x/DLL/Packet32.c#L134-L180
 * https://msdn.microsoft.com/en-us/library/windows/desktop/aa363216(v=vs.85).aspx

It's worth logging each of the parameters before calling `PacketReceivePacket()` - that way, when it fails, you can see exactly what input caused it to fail.

**Comment by bgourlie** [comment]
A few questions (not apparent reading the linked docs):

- Is it really exceptional for ` winpcap::PacketReceivePacket` to return 0?
- If not, shouldn't the iterator that's being implemented return an `Option`

**Comment by mrmonday** [comment]
According to the docts there:

```
If the operation fails or is pending, the return value is zero. 
```

So I believe that the error there is correct (except possibly in the pending case - that needs more investigation - might be worth opening a separate issue for that).

For the original error you reported to occur, `PacketReceivePacket()` needs to return success (non-zero), but after that no packets get added to `self.packets`.

**Comment by mrmonday** [comment]
Have you had a chance to look into this further?

**Comment by micmonay** [comment]
I have created a solution for me. I returned an error instead of unwrap. For me it's good, but I don't know if my code is correctly written. [292154d](https://github.com/micmonay/libpnet/commit/292154d1ae38cd84d21d78e3f6068d85efcc0c37)

**Comment by mrmonday** [comment]
@micmonay Are you able to reliably reproduce the issue?

Looking at the fix, I'd be curious to see what the value of `buflen` is when this fails. Perhaps we need an additional check in there?

Either way, I'd love it if you could make a PR for this!

**Comment by micmonay** [comment]
I have changed code for show debug information for buflen
```rust
impl DataLinkReceiver for DataLinkReceiverImpl {
    fn next(&mut self) -> io::Result<&[u8]> {
        let mut debug = String::new(); // <-- new string for debug
        if self.packets.is_empty() {
            let ret = unsafe {
                winpcap::PacketReceivePacket(self.adapter.adapter, self.packet.packet, 0)
            };
            let buflen = match ret {
                0 => return Err(io::Error::last_os_error()),
                _ => unsafe { (*self.packet.packet).ulBytesReceived as isize },
            };
            let mut ptr = unsafe { (*self.packet.packet).Buffer };
            let end = unsafe { (*self.packet.packet).Buffer.offset(buflen) };
            while ptr < end {
                unsafe {
                    let packet: *const bpf::bpf_hdr = mem::transmute(ptr);
                    let start = ptr as isize + (*packet).bh_hdrlen as isize -
                                (*self.packet.packet).Buffer as isize;
                    self.packets.push_back((start as usize, (*packet).bh_caplen as usize));
                    let offset = (*packet).bh_hdrlen as isize + (*packet).bh_caplen as isize;
                    ptr = ptr.offset(bpf::BPF_WORDALIGN(offset));
                }
            }
            debug.push_str(&format!("size buflen {}",buflen));// <-- append debug information for buflen
        }
        let (start, len) = match self.packets.pop_front(){
            Some(p) => (p),
            None => {
                        // Show if packets is empty
                return Err(Error::new(ErrorKind::Interrupted,format!("packets is empty trace :\n{}",debug.to_string())));  
                }
        };
        let slice = unsafe {
            let data = (*self.packet.packet).Buffer as usize + start;
            slice::from_raw_parts(data as *const u8, len)
        };
        Ok(slice)
    }
}
```
After few minute I get this result :
```
packets is empty trace :
size buflen 0
packets is empty trace :
size buflen 0
packets is empty trace :
size buflen 0
packets is empty trace :
size buflen 0
packets is empty trace :
size buflen 0
packets is empty trace :
size buflen 0
...
```
This don't breaking the capture
On the next result, I capture the Homeplug packets:
```
ff:ff:ff:ff:ff:ff ethertype: EtherType(35041) length: 60
ff:ff:ff:ff:ff:ff ethertype: EtherType(35041) length: 60
ff:ff:ff:ff:ff:ff ethertype: EtherType(35041) length: 60
ff:ff:ff:ff:ff:ff ethertype: EtherType(35041) length: 60
packets is empty trace :
size buflen 0
packets is empty trace :
size buflen 0
packets is empty trace :
size buflen 0
... (more lines)
ff:ff:ff:ff:ff:ff ethertype: EtherType(35041) length: 60
packets is empty trace :
size buflen 0
packets is empty trace :
size buflen 0
... (more lines)
ff:ff:ff:ff:ff:ff ethertype: EtherType(35041) length: 60
```
I'm not sure if I have understand your request. If you have need other test...

**Comment by mrmonday** [comment]
@micmonay Thanks for the debugging!

The issue is that `PacketReceivePacket` is returning success, but not returning a packet - the code doesn't account for this.

Given that, I think  the correct fix is to change the `if self.packets.is_empty()` to `while self.packets.is_empty()`, then check for `buflen == 0` - if it is, continue to the next iteration (aka try and get another packet).

**Comment by micmonay** [comment]
Just `while` it's fine for me in my test

**Comment by mrmonday** [comment]
Going to close this, since it should be fixed by #327. If there's still an issue this can be re-opened.

Thanks again @micmonay!
